design/BufferOperations.md

```diff
@@ -0,0 +1,147 @@
+# Buffer operations
+
+This explainer describes the operations that are available on the `WebGPUBuffer` object directly.
+They are `setSubData` which is an immediate data upload operations, and `mapWriteAsync`, `mapReadAsync` and `unmap` which are memory mapping operations.
+The explainer also defines `WebGPUMappedMemory` objects which are used for buffer mapping.
+
+## Prelimineries: buffered / unbuffered commands
+
+Assuming there is a single queue, there are two types of commands in WebGPU:
+
+ - "Buffered commands": commands stored inside `WebGPUCommandBuffer` like `setVertexBuffers` or `draw`.
+ - "Unbuffered commands": Other commands like creating objects, `WebGPUBuffer.setSubData` and `WebGPUQueue.submit`.
+
+Assuming there is a single queue, there is a total order on the unbuffered commands: they all execute atomically in the order they were called.
+`WebGPUQueue.submit` is special because it atomically executes all the commands stored in its `commands` argument.
+
+## `WebGPUBuffer.setSubData`
```

Comment:
Seems like @Kangz already decided he likes setSubData better.  :-)

Reply:
@kainino0x you're right that `mapWriteSync` avoids a copy at the cost of adding a GC object.

I like `mapWriteSync` better and would even be happier with both options, but my feeling from last meeting was that the grouped was happier with `setSubData`. I might be wrong.

Reply:
Firstly, I don't like `setSubData` being Unbuffered, in Vulkan these sorts of updates are buffered so its possible to update data between render-passes without having to split up command buffers.


Secondly but more importantly, `setSubData` may turn out to be more important and useful than `mapWriteAsync` for zero-copy resource intialization. 

The use cases are raytracing and Deferred+ when descriptor indexing is not available, when we need to pack all of our intersection-kernel accesible data into one GPU buffer/descriptor. That buffer could not be duplicated or round-robined because of its sheer size (all of the scene geometry) so mapping could not be used to update it (unless a separate staging buffer is used).

Obviously #156 solves that problem.

Reply:
If the application wishes to update the content of a buffer in the middle of a command buffer, it can do so using `copyBufferToBuffer` and having previously staged the data for example using the RingBufferOfBuffer.js approach. Some hardware is able to inline "setSubData" data in the command buffer, but in most cases I expect Vulkan implementation are just managing staging memory so the result is equivalent to what I described above.

I believe this also addresses the concern for raytracing and Deferred+.

Really `setSubData` should be used for only two cases:
 - You are new to WebGPU and the tutorial doesn't want to explain buffer mapping just yet.
 - You want to debug GPU code and need to fill in a buffer with data right now.

Reply:
> I believe this also addresses the concern for raytracing and Deferred+

Well, the whole point of my comment was to highlight that you're not get zero-copy initialisation ("unless a separate staging buffer is used").

On UMA systems with persistent mapping one could zero-copy init the resources for Deferred+ and raytracing.

Reply:
Yes and we agree #156 would help with this. I just don't think it informs the issue about buffered vs. unbuffered `setSubData`.

Reply:
But again, #156 buffer-sub-ranges (views) would make zero-init possible with mapping